File: docs/snippets/typescript/xml-evaluation-runner_main.ts
import { AgentProtocolClient } from '@microsoft/agents-protocol-client';
import { MessageDeserializer } from '@microsoft/agents-xml';
import { ThreadValidator } from '@microsoft/agents-validation';
import { ChatMessage, TextContent } from '@microsoft/agents';
import { promises as fs } from 'fs';
import { glob } from 'glob';

class EvaluationRunner {
  private client: AgentProtocolClient;
  private deserializer = new MessageDeserializer();
  private validator = new ThreadValidator();

  constructor(agentEndpoint: string) {
    this.client = new AgentProtocolClient({ baseUrl: agentEndpoint });
  }

  async runEvaluation(testFilePath: string): Promise<boolean> {
    try {
      // Load evaluation test case from XML
      const evalXml = await fs.readFile(testFilePath, 'utf-8');
      const testMessages = this.deserializer.deserializeMany(evalXml);

      // Extract the user input from the test case using type guard
      const userMessage = testMessages.find((m): m is ChatMessage & { role: 'user' } => m.role === 'user');
      if (!userMessage) {
        throw new Error('No user message found in test case');
      }

      const expectedMessages = testMessages.filter(m => m.role !== 'user');

      // Send to agent via Client SDK
      const conversation = this.client.createConversation();

      // Use type guard for text content
      const textContent = userMessage.contents.find((c): c is TextContent => c.kind === 'text');
      if (!textContent) {
        throw new Error('No text content found in user message');
      }

      await conversation.send(textContent.text);

      // Get messages from local cache (no HTTP call)
      const actualMessages = conversation.messages;

      // Validate actual vs expected behavior
      const validationResult = this.validator.validate(actualMessages, expectedMessages);

      if (validationResult.isValid) {
        console.log(`✓ Test passed: ${testFilePath}`);
        return true;
      } else {
        console.log(`✗ Test failed: ${testFilePath}`);
        validationResult.errors.forEach(error => {
          console.log(`  - ${error.message}`);
        });
        return false;
      }
    } catch (error) {
      console.error(`Evaluation failed: ${testFilePath}`, error);
      return false;
    }
  }
}

// Usage: Run all evals
const runner = new EvaluationRunner("http://localhost:5000");
const testFiles = await glob("test-cases/*.xml");
for (const testFile of testFiles) {
  await runner.runEvaluation(testFile);
}

File: docs/snippets/python/xml-evaluation-runner_main.py
import asyncio
from typing import List
from pathlib import Path
from microsoft.agents.protocol import AgentProtocolClient
from microsoft.agents.xml import XmlDeserializer
from microsoft.agents.validation import ThreadValidator
from microsoft.agents.models import UserMessage, TextContent, ChatMessage


class EvaluationRunner:
    def __init__(self, agent_endpoint: str):
        self.client = AgentProtocolClient(agent_endpoint)
        self.deserializer = XmlDeserializer()
        self.validator = ThreadValidator()

    async def run_evaluation(self, test_file_path: str) -> bool:
        """Run evaluation test case against live agent."""
        # Load evaluation test case from XML
        test_xml = Path(test_file_path).read_text()
        test_messages: List[ChatMessage] = self.deserializer.deserialize_many(test_xml, root_element="eval")

        # Extract the user input from the test case
        user_message = next(m for m in test_messages if isinstance(m, UserMessage))
        expected_messages = [m for m in test_messages if not isinstance(m, UserMessage)]

        # Send to agent via Client SDK
        conversation = self.client.create_conversation()
        user_text = next(c for c in user_message.contents if isinstance(c, TextContent)).text
        await conversation.send(user_text)

        # Get messages from local cache (no HTTP call)
        actual_messages = conversation.messages

        # Validate actual vs expected behavior
        validation_result = self.validator.validate(actual_messages, expected_messages)

        if validation_result.is_valid:
            print(f"✓ Test passed: {test_file_path}")
            return True
        else:
            print(f"✗ Test failed: {test_file_path}")
            for error in validation_result.errors:
                print(f"  - {error.message}")
            return False


async def main() -> None:
    # Usage: Run all evals
    runner = EvaluationRunner("http://localhost:5000")
    test_files = Path("test-cases").glob("*.xml")
    for test_file in test_files:
        await runner.run_evaluation(str(test_file))


if __name__ == "__main__":
    asyncio.run(main())
